### PR TITLE
Cereal serialization - Update serialization in Any.h

### PR DESCRIPTION
**Pull request by OXPHOS**
It's super verbose because - 
1. Need to save the type information by casting the types of values to Enum class
2. Need to extract the type information when saving/loading.
The second part is a little bit better because I reshaped with macros. 
But I don't have any better idea for the first part.
I think `SGMatrix` is the only thing missed here.
I'll add more tests once it looks good.

@vigsterkr @karlnapf

### CODE REVIEW

**Comment by karlnapf** [comment]
I think this branch needs a rebase (against what?)
check out the commit about googletest that appeared

File: src/shogun/lib/any.h [hunk 1]
```diff
@@ -45,21 +45,58 @@ namespace shogun
...
+
+namespace serial
+{
 		enum EnumContainerType
 		{
 			CT_UNDEFINED,
 			CT_PRIMITIVE,
-			CT_SGVECTOR,
-			CT_SGMATRIX
+			CT_SGVector,
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: why this?

File: src/shogun/lib/any.h [hunk 2]
```diff
@@ -45,21 +45,58 @@ namespace shogun
...
+			CT_SGMatrix
 		};
 
 		enum EnumPrimitiveType
 		{
 			PT_UNDEFINED,
-			PT_INT_32,
-			PT_FLOAT_64,
+			PT_bool,
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: why do you rename?

File: src/shogun/lib/any.h [hunk 3]
```diff
@@ -312,31 +525,49 @@ namespace shogun
...
-					break;
-				case serial::EnumPrimitiveType::PT_FLOAT_64:
-					cereal_save_helper<Archive, float64_t>(ar);
+				CASE_PRIMITIVE_SAVE(bool);
+				CASE_PRIMITIVE_SAVE(char);
+				CASE_PRIMITIVE_SAVE(int8_t);
+				CASE_PRIMITIVE_SAVE(uint8_t);
+				CASE_PRIMITIVE_SAVE(int16_t);
+				CASE_PRIMITIVE_SAVE(uint16_t);
+				CASE_PRIMITIVE_SAVE(int32_t);
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: seems ok for me to do it this way

File: src/shogun/lib/any.h [hunk 4]
```diff
@@ -370,37 +612,55 @@ namespace shogun
...
+				CASE_SGOBJECT_LOAD(SGVector, complex128_t);
 				case serial::EnumPrimitiveType::PT_UNDEFINED:
 					SG_SERROR("Type error: undefined data type cannot be serialized.\n");
 					break;
 				}
 				break;
 
-			case serial::EnumContainerType::CT_SGMATRIX:
+			case serial::EnumContainerType::CT_SGMatrix:
 				SG_SWARNING("SGMatrix serializatino method not implemented.\n")
```
Comment (karlnapf) [inline src/shogun/lib/any.h]: typo in warning message

**Comment by OXPHOS** [comment]
I used lowercase to fit the [macros](https://github.com/shogun-toolbox/shogun/pull/3696/files#diff-8ea96286d95b52029d31636117e0fe55R50)
I'll rebase the branch. However wiking told me another project did it pretty concise so I was going to check that out.

**Comment by karlnapf** [comment]
Shall we re-vive this?

**Comment by gf712** [comment]
@vigsterkr I'm guessing this can be closed too then?

**Comment by vigsterkr** [comment]
ciao bella!

**Pull request closed without merging**